rust/wasmtime_bindings/src/lib.rs

```diff
@@ -0,0 +1,347 @@
+}
+
+fn create_store(engine: &mut Engine) -> Result<Box<Store>> {
+    let wasi = wasmtime_wasi::WasiCtxBuilder::new().build();
+    let mut store = wasmtime::Store::new(&engine.wasmtime_engine, wasi);
+    // TODO: make this configurable
+    store.out_of_fuel_async_yield(10000000, 1000);
+    Ok(Box::new(Store {
+        wasmtime_store: store,
+    }))
+}
+
+pub struct Memory {
+    wasmtime_memory: wasmtime::Memory,
+}
+
+impl Memory {
+    fn data(self: &Memory, store: &Store) -> *mut u8 {
+        self.wasmtime_memory.data_ptr(&store.wasmtime_store)
+    }
+    fn size(self: &Memory, store: &Store) -> u64 {
+        self.wasmtime_memory.size(&store.wasmtime_store)
+    }
+    fn grow(self: &mut Memory, store: &mut Store, delta: u64) -> Result<u64> {
+        self.wasmtime_memory.grow(&mut store.wasmtime_store, delta)
+    }
+}
+
+fn get_memory(instance: &Instance, store: &mut Store) -> Result<Box<Memory>> {
+    let export = instance
+        .wasmtime_instance
+        .get_export(&mut store.wasmtime_store, "memory")
+        .ok_or_else(|| {
+            anyhow!("Memory export not found - please export `memory` in the wasm module")
+        })?;
+    let memory = export
+        .into_memory()
+        .ok_or_else(|| anyhow!("Exported object memory is not a WebAssembly memory"))?;
+    Ok(Box::new(Memory {
+        wasmtime_memory: memory,
+    }))
+}
+
+fn get_abi(instance: &Instance, store: &mut Store, memory: &Memory) -> Result<u32> {
+    let export = instance
+        .wasmtime_instance
+        .get_export(&mut store.wasmtime_store, "_scylla_abi")
+        .ok_or_else(|| {
+            anyhow!("ABI export not found - please export `_scylla_abi` in the wasm module")
+        })?;
+    let global = export
+        .into_global()
+        .ok_or_else(|| anyhow!("Exported object _scylla_abi is not a WebAssembly global"))?;
+    if let wasmtime::Val::I32(x) = global.get(&mut store.wasmtime_store) {
+        let mut bytes = [0u8; 4];
+        memory
+            .wasmtime_memory
+            .read(&store.wasmtime_store, x as usize, &mut bytes)?;
+        Ok(u32::from_le_bytes(bytes))
+    } else {
+        Err(anyhow!("Exported global _scylla_abi is not an i32"))
+    }
+}
+
+pub struct Engine {
+    wasmtime_engine: wasmtime::Engine,
+}
+
+fn create_engine() -> Result<Box<Engine>> {
+    let mut config = wasmtime::Config::new();
+    config.async_support(true);
+    config.consume_fuel(true);
+    // Parallel compilation is enabled by default, but we can't use it in a seastar context.
+    config.parallel_compilation(false);
+    // ScyllaMemoryCreator uses malloc (from seastar) to allocate linear memory
+    config.with_host_memory(std::sync::Arc::new(ScyllaMemoryCreator {}));
+    // The following configuration settings make wasmtime allocate only as much memory as it needs
+    config.static_memory_maximum_size(0);
+    config.dynamic_memory_reserved_for_growth(0);
+    config.dynamic_memory_guard_size(0);
```

Comment:
Without the guard, will it emit bound checks?

Reply:
Yes. If we didn't provide our own memory allocator, the memory would be allocated using mmap by default and the guard would actually allow eliding some of the checks - the guard region would remain unmapped and would trigger a segfault if accessed.